src/Interpreters/InterpreterSystemQuery.cpp

```diff
@@ -373,10 +374,17 @@ BlockIO InterpreterSystemQuery::execute()
             break;
         }
         case Type::RELOAD_MODEL:
+        {
+            getContext()->checkAccess(AccessType::SYSTEM_RELOAD_MODEL);
+            auto bridge_helper = std::make_unique<CatBoostLibraryBridgeHelper>(getContext(), query.target_model);
+            bridge_helper->removeModel();
```

Comment:
why not actually reload and not just remove? I would expect that time for reloading the model lies on the reload model command instead of the next attempt to access it

Reply:
I agree in general (even if the reload costs are small because models are small)

The problem is that the model evaluation is not an atomic operation. It is actually two operations A and B with a dependency between them. This cannot be changed easily because of ClickHouse's function interface.

So we need to deal with a `SYSTEM RELOAD MODEL` call between A and B. The simplest approach is the one implemented: Simply unload the model, and let B check if the model is still in memory. If A finds a model unloaded, load it.

If `SYSTEM RELOAD MODEL` actually *reloads* the model, then B must somehow find that out and fail. There are multiple approaches. I'll mention just two
- A locks the model on the bridge side until B starts. This avoids reloads in between. Obviously there is a risk of livelocks here ...
- More sophisticated: Generated for each `catboostEvaluate()` call a random "request id" on the server side, send it to the bridge in A and store it in a queue. A reload in between clears the queue. The same request id is send to the bridge in B. If it is not found in the queue, fail. The length of the queue limits the number of parallel requests which is why I did not like this option too much. It's also harder to maintain and test.

TLDL: I think the current behavior is not a deal breaker for such quite exotic functionality. If it's okay, I'll ask Nikolai what he thinks (he was the original author) and if needed implement an actual load in a separate PR.

Reply:
For completeness, I talked briefly to Nikolai yesterday and he's fine with the lazy semantics.